test/k8s_service_example_test_helpers.go

```diff
@@ -176,3 +182,58 @@ func verifyServiceAvailable(
 		validationFunction,
 	)
 }
+
+// verifyServiceRoutesToMainAndCanaryPods ensures that the service is routing to both the main and the canary pods
+// It does this by repeatedly issuing requests to the service and inspecting the nginx Server header
+// Once both nginx tags have been seen in this header - we can be confident that we've reached both types of pod via the service
+func verifyServiceRoutesToMainAndCanaryPods(
+	t *testing.T,
+	kubectlOptions *k8s.KubectlOptions,
+	appName string,
+	releaseName string,
+) {
+
+	// Get the service and wait until it is available
+	filters := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("app.kubernetes.io/name=%s,app.kubernetes.io/instance=%s", appName, releaseName),
+	}
+	services := k8s.ListServices(t, kubectlOptions, filters)
+	require.Equal(t, len(services), 1)
+	service := services[0]
+	k8s.WaitUntilServiceAvailable(t, kubectlOptions, service.Name, WaitTimerRetries, WaitTimerSleep)
+
+	service = *k8s.GetService(t, kubectlOptions, service.Name)
+	serviceEndpoint := k8s.GetServiceEndpoint(t, kubectlOptions, &service, 80)
+
+	// Ensure that the service routes to both the main and canary deployment pods
+	// Read the latest values dynamically in case the fixtures file changes
+	valuesFile, err := ioutil.ReadFile(filepath.Join("fixtures", "canary_and_main_deployment_values.yaml"))
+	assert.NoError(t, err)
+
+	rendered := map[string]interface{}{}
+	err = yaml.Unmarshal([]byte(valuesFile), &rendered)
+
+	mainImageTag := rendered["containerImage"].(map[string]interface{})["tag"].(string)
+
+	canaryImageTag := rendered["canary"].(map[string]interface{})["containerImage"].(map[string]interface{})["tag"].(string)
+
+	// We haven't seen either tag come back in the nginx Server header yet
+	seen := make(map[string]bool)
+	seen[mainImageTag] = false
+	seen[canaryImageTag] = false
+
+	for seen[mainImageTag] == false || seen[canaryImageTag] == false {
```

Comment:
This infinite loop is a bit dangerous if you have a test failure where only one of the Pod types come up, because you will never see the canary. You want some kind of timeout logic here or max retries.

You can use [`retry.DoWithRetry` from terratest](https://github.com/gruntwork-io/terratest/blob/master/modules/retry/retry.go#L53) for this purpose. See https://github.com/gruntwork-io/terratest/blob/master/test/terraform_ssh_example_test.go#L115 for example usage.

Reply:
Understood - will fix

Reply:
I wrapped the tag check loop in `retry.DoWithRetry` in 838299c. In my local testing it works well, and both tags are seen before the max retries value of 30 is hit. 

Let me know what you think about that value.